source/extensions/io_socket/user_space/io_handle_impl.h

```diff
@@ -0,0 +1,184 @@
+  Network::Address::InstanceConstSharedPtr peerAddress() override;
+
+  void initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                           Event::FileTriggerType trigger, uint32_t events) override;
+  Network::IoHandlePtr duplicate() override;
+  void activateFileEvents(uint32_t events) override;
+  void enableFileEvents(uint32_t events) override;
+  void resetFileEvents() override;
+
+  Api::SysCallIntResult shutdown(int how) override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override { return absl::nullopt; }
+
+  void setWatermarks(uint32_t watermark) { pending_received_data_.setWatermarks(watermark); }
+  void onBelowLowWatermark() {
+    if (peer_handle_) {
+      ENVOY_LOG(debug, "Socket {} switches to low watermark. Notify {}.", static_cast<void*>(this),
+                static_cast<void*>(peer_handle_));
+      peer_handle_->onPeerBufferLowWatermark();
+    }
+  }
+  void onAboveHighWatermark() {
+    // Low to high is checked by peer after peer writes data.
+  }
+
+  // UserSpace::IoHandle
+  void setWriteEnd() override {
+    receive_data_end_stream_ = true;
+    setNewDataAvailable();
+  }
+  void setNewDataAvailable() override {
+    ENVOY_LOG(trace, "{} on socket {}", __FUNCTION__, static_cast<void*>(this));
+    if (user_file_event_) {
+      user_file_event_->activateIfEnabled(
+          Event::FileReadyType::Read |
+          // Closed ready type is defined as `end of stream`
+          (receive_data_end_stream_ ? Event::FileReadyType::Closed : 0));
+    }
+  }
+  void onPeerDestroy() override {
+    peer_handle_ = nullptr;
+    write_shutdown_ = true;
+  }
+  void onPeerBufferLowWatermark() override {
+    if (user_file_event_) {
+      user_file_event_->activateIfEnabled(Event::FileReadyType::Write);
+    }
+  }
+  bool isWritable() const override { return !pending_received_data_.highWatermarkTriggered(); }
+  bool isPeerShutDownWrite() const override { return receive_data_end_stream_; }
+  bool isPeerWritable() const override {
+    return peer_handle_ != nullptr && !peer_handle_->isPeerShutDownWrite() &&
+           peer_handle_->isWritable();
+  }
+  Buffer::Instance* getWriteBuffer() override { return &pending_received_data_; }
+
+  // `UserspaceIoHandle`
+  bool isReadable() const override {
+    return isPeerShutDownWrite() || pending_received_data_.length() > 0;
+  }
+
+  // Set the peer which will populate the owned pending_received_data.
+  void setPeerHandle(UserSpace::IoHandle* writable_peer) {
+    // Swapping writable peer is undefined behavior.
+    ASSERT(!peer_handle_);
+    ASSERT(!write_shutdown_);
+    peer_handle_ = writable_peer;
+  }
+
+private:
+  friend class IoHandleFactory;
+  IoHandleImpl();
+
+  // Support isOpen() and close(). Network::IoHandle owner must invoke close() to avoid potential
+  // resource leak.
+  bool closed_{false};
+
+  // The attached file event with this socket. The event is not owned by the socket in the current
```

Comment:
This comment seems at odds with the field being an unique_ptr which implies ownership.

Reply:
stale comment